<a href="https://colab.research.google.com/github/MiguelDz89/MIAR_Algoritmos_Optimizacion/blob/main/Seminario_Algoritmos_Miguel_Diaz.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Algoritmos de optimización - Seminario<br>
Nombre y Apellidos: **Miguel Angel Díaz Arica**  <br>
Url: https://github.com/MiguelDz89/MIAR_Algoritmos_Optimizacion/blob/main/Seminario_Algoritmos_Miguel_Diaz.ipynb<br>
Problema:
> 1. Sesiones de doblaje <br>
>2. Organizar los horarios de partidos de La Liga<br>
>3. Combinar cifras y operaciones

Descripción del problema:

# **Combinar cifras y operaciones**

• El problema consiste en analizar el siguiente problema y diseñar un algoritmo que lo resuelva.

• Disponemos de las 9 cifras del 1 al 9 (excluimos el cero) y de los 4 signos básicos de las
operaciones fundamentales: suma(+), resta(-), multiplicación(*) y división(/)

• Debemos combinarlos alternativamente sin repetir ninguno de ellos para obtener una
cantidad dada. Un ejemplo sería para obtener el 4:
4+2-6/3*1 = 4

Debe analizarse el problema para encontrar todos los valores enteros posibles planteando las siguientes cuestiones:
- ¿Qué valor máximo y mínimo se pueden obtener según las condiciones del problema?

- ¿Es posible encontrar todos los valores enteros posibles entre dicho mínimo y máximo ?

• Nota: Es posible usar la función de python “eval” para evaluar una expresión, ejemplo:
print(eval('4+2-6/3*1'))
resultado=4


(*) La respuesta es obligatoria





                                        

## Pseudocódigo del Algoritmo
Se detalla el pseudocódigo de lo que se va a programar, hay que tener en cuenta que vamos a recorrer un listado con las cifras y un listado de los operadores ambos los generamos por permutacion:
```
generar todas las permutaciones de las cifras del 1 al 9
generar todas las permutaciones de los operadores +, -, *, /

para cada permutación de cifras:
    para cada permutación de operadores:
        combinar cifras y operadores para formar una expresión
        evaluar la expresión
        si el resultado es un entero:
            almacenar el resultado en un conjunto

determinar el valor mínimo de los resultados enteros
determinar el valor máximo de los resultados enteros

verificar si todos los valores enteros entre el mínimo y el máximo están presentes

imprimir el valor mínimo, valor máximo y si todos los valores están presentes

```




In [6]:
import itertools

# Generar todas las permutaciones de las cifras del 1 al 9
digits = list(map(str, range(1, 10)))
digit_permutations = list(itertools.permutations(digits))

# Generar todas las combinaciones de los 4 operadores
operators = ['+', '-', '*', '/']
operator_permutations = list(itertools.permutations(operators))

list_expr = {}

# Función para intercalar cifras y operadores
def intercalar(cifras, operadores):
    expresion = ''
    for i in range(len(operadores)):
        expresion += cifras[i] + operadores[i]
    expresion += cifras[-1]
    return expresion

# Evaluar todas las expresiones posibles
resultados = set()
for dp in digit_permutations:
    for op in operator_permutations:
        try:
            expr = intercalar(dp, op)
            resultado = eval(expr)
            if resultado == int(resultado):  # Considerar solo resultados enteros
                resultados.add(int(resultado))
                list_expr[expr]=resultado
        except ZeroDivisionError:
            continue

# Determinar valores máximos y mínimos
if resultados:
    valor_minimo = min(resultados)
    valor_maximo = max(resultados)
    # Verificar si todos los valores entre mínimo y máximo están presentes
    todos_valores_presentes = all(x in resultados for x in range(valor_minimo, valor_maximo + 1))
else:
    valor_minimo = None
    valor_maximo = None
    todos_valores_presentes = False

# Resultados

print("Cantidad de digit_permutations:", len(digit_permutations))
print("Cantidad de operator_permutations:", len(operator_permutations))
print("Numero de expresiones generadas:", len(list_expr))
print("Valor mínimo:", valor_minimo)
print("Valor máximo:", valor_maximo)
print("Todos los valores enteros entre mínimo y máximo están presentes:", todos_valores_presentes)


Valor digit_permutations: 362880
Valor operator_permutations: 24
Numero de expresiones generadas: 90000
Valor mínimo: -69
Valor máximo: 77
Todos los valores enteros entre mínimo y máximo están presentes: True


# ¿Cuantas posibilidades hay sin tener en cuenta las restricciones?<br>



Para tener en cuenta primero debemos calcular:

* **Permutaciones de las cifras <br>**
  Hay 9 cifras diferentes del 1 al 9.<br>
  Las permutaciones de 9 se calculan como 9 factorial(9!) = 362880.<br>

* **Permutaciones de los operadores**<br>
  Hay 4 operadores diferentes (+, -, *, /).<br>
  Las permutaciones de 4 elementos se calculan como 4 factorial(4!) = 24.<br>

* **Calcular las combinaciones posibles**
  Cada permutación de cifras lo combinaremos con cada permutación de operadores, siendo el número total de combinaciones posibles sin restricciones: 9! x 4! = 362880 x 24 = **8 718 720**.

# ¿Cuantas posibilidades hay teniendo en cuenta todas las restricciones?


Para tener en cuenta primero debemos calcular:

* **Permutaciones de las cifras <br>**
  Hay 9 cifras diferentes del 1 al 9.<br>
  Las permutaciones de 9 se calculan como 9 factorial(9!) = 362880.<br>

* **Permutaciones de los operadores**<br>
  Hay 4 operadores diferentes (+, -, *, /).<br>
  Las permutaciones de 4 elementos se calculan como 4 factorial(4!) = 24.<br>

* **Intercalar cifras y operaciones**
  Por cada permutación de cifras, hay 24 formas de intercalar los operadores.

  Hay que tener en cuenta que no todas las combinaciones nos darán un resultado de números enteros, por lo que debemos filtrar los resultados que no cumplen con esta condición.
  
  Evaluando el código obtenemos una evaluación de **80000** combinaciones válidas



Modelo para el espacio de soluciones<br>
# (*) ¿Cual es la estructura de datos que mejor se adapta al problema? Argumentalo.(Es posible que hayas elegido una al principio y veas la necesidad de cambiar, arguentalo)

Para resolver este problema, es importante seleccionar estructuras de datos que optimicen tanto la generación de combinaciones como la evaluación y almacenamiento de resultados. Evaluemos las estructuras de datos utilizadas en el código anterior:

**Estructuras de datos utilizadas:**
* Lista (list) para almacenar permutaciones de cifras y operadores:<br>
Las listas son adecuadas para almacenar permutaciones generadas por itertools.permutations debido a su facilidad de acceso y manipulación indexada.<br>
La ventaja de usar listas es por su fácil acceso por índice, simplicidad en la construcción y lectura, soporta iteración eficiente.<br>
Una desventaja que podemos encontrar es que no garantiza unicidad de elementos y puede tener problemas de rendimiento para operaciones frecuentes de búsqueda y eliminación en listas grandes.
* Conjunto (set) para almacenar resultados enteros:<br>
Son ideales para almacenar resultados únicos debido a su propiedad de no permitir duplicados y ofrecer operaciones de búsqueda rápidas. Esto nos permite realizar operaciones de inserción, eliminación y búsqueda con complejidad promedio O(1) que nos garantiza unicidad de elementos.

Según el modelo para el espacio de soluciones<br>
# (*)¿Cual es la función objetivo?
La función objetivo la podemos definir a partir del problema que queremos optimizar.
## Función objetivo:
La función objetivo en este problema sería evaluar todas las expresiones posibles generadas por la intercalación de las permutaciones de las cifras y operadores, y contar cuántas de estas evaluaciones resultan en números enteros.



# (*)¿Es un problema de maximización o minimización?

El problema es principalmente de maximización. Queremos maximizar el conjunto de resultados enteros posibles, es decir, queremos identificar la mayor cantidad de expresiones diferentes que resulten en números enteros. También podríamos considerar maximizar el rango de valores enteros obtenidos (es decir, encontrar el mayor y el menor valor entero posible).

# Diseña un algoritmo para resolver el problema por fuerza bruta

Respuesta: <br>
El algoritmo diseñado se realizo utilizando la fuerza bruta.

# Calcula la complejidad del algoritmo por fuerza bruta

Respuesta: <br>
Para calcular la complejidad del algoritmo de fuerza bruta que se utiliza en este problema, consideremos los diferentes pasos del algoritmo y las operaciones involucradas:

1. Generación de permutaciones de las cifras: <br>
* Hay 9 cifras (del 1 al 9).
* El número de permutaciones posibles de 9 elementos es 9!.
* La complejidad de generar todas las permutaciones de las cifras es O(9!).
2. Generación de permutaciones de los operadores:
* Hay 4 operadores (+, -, *, /).
* El número de permutaciones posibles de 4 elementos es 4!.
* La complejidad de generar todas las permutaciones de los operadores es O(4!).
3. Combinación de permutaciones de cifras y operadores:
* Para cada permutación de cifras, generamos todas las permutaciones de los operadores.
* El número total de combinaciones de cifras y operadores es 9!×4!.
4. Evaluación de expresiones: <br>
* Para cada combinación de cifras y operadores, se forma y evalúa una expresión.
* La evaluación de una expresión tiene una complejidad O(1).
* Por lo tanto, la complejidad de evaluar todas las combinaciones es O(9!×4!).
## Complejidad total del algoritmo por fuerza bruta:
Dado que las operaciones de generación de permutaciones y evaluación de expresiones dominan el tiempo de ejecución del algoritmo, la complejidad total del algoritmo es: O(9!×4!)

**Calculando los valores:** <br>
* 9!=362880 <br>
* 4!=24 <br>
Entonces, el número total de combinaciones posibles es:
9!×4! = 362880×24 = **8718720** <br>

La complejidad del algoritmo es **O(8718720)**, que se puede simplificar a **O(9!×4!)**.

# (*)Diseña un algoritmo que mejore la complejidad del algortimo por fuerza bruta. Argumenta porque crees que mejora el algoritmo por fuerza bruta

## Enfriamiento simulado, Recocido Simulado,  o Simulated-Annealing
El Enfriamiento Simulado o Recocido Recocido es un método de optimización que permite movimientos a soluciones peores con una probabilidad que disminuye con el tiempo (temperatura). La probabilidad de aceptar una solución peor se basa en la función de temperatura que decrece a lo largo del tiempo.
Este tipo de algoritmos tiene como objetivo desarrollar técnicas eficientes para encontrar valores mínimos y máximos de una función de muchas variables independientes. La función de costos depende de la configuración detallada del sistema.

1. En este caso vamos a inicializar el problema definiendo los parametros iniciales, temperatura inicial, temperatura final, tasa de enfriamiento y número de iteraciones por temperatura. Luego generamos una solución inicial y calculamos su fitness. <br>
2. Luego inicializaremos la temperatura y calculamos el número de iteraciones que necesitaremos para que la temperatura disminuya desde la temperatura Inicial hasta la final usando la tasa de enfriamiento α.<br>
3. Creamos un bucle de enfriamiento donde la condición será que mientras la temperatura no sea menor que la Temperatura final se realizan las números de evaluaciones por iteración(ITERATIONS_PER_TEMP):
* Cada evaluación genera un vecino y se evalúa su fitness.
* Se decide si se acepta el vecino basado en la diferencia de fitness y la temperatura actual.
* Se evalúa la expresión de la solución actual y se almacena si el resultado es un entero.
* La temperatura se reduce multiplicándola por la tasa de enfriamiento
α.
<br>

4. Al finalizar el bucle de enfriamiento, se evalúan los resultados finales y se determinan los valores mínimo y máximo y sus expresiones correspondientes.



## Porque mejora el algoritmo de fuerza bruta:
El algoritmo de enfriamiento simulado mejora el enfoque de fuerza bruta principalmente debido a su capacidad para explorar el espacio de soluciones de manera más eficiente y efectiva. Reduce significativamente el número de evaluaciones necesarias y utiliza una estrategia probabilística para evitar quedar atrapado en óptimos locales. Esta combinación de factores lo hace mucho más adecuado para problemas de gran escala donde la fuerza bruta sería impracticable debido a su alta complejidad computacional.

In [5]:
import random
import math

# Parámetros del algoritmo de enfriamiento simulado
INITIAL_TEMPERATURE = 10000
FINAL_TEMPERATURE = 1
ALPHA = 0.995  # Tasa de enfriamiento
ITERATIONS_PER_TEMP = 100

# Generar una expresión a partir de dígitos y operadores
def generar_expresion(digitos, operadores):
    expresion = ''
    for i in range(len(operadores)):
        expresion += digitos[i] + operadores[i]
    expresion += digitos[-1]
    return expresion

# Función para evaluar una expresión y verificar si es un entero
def evaluar_expresion(expr):
    try:
        resultado = eval(expr)  # Evalúa la expresión matemática
        if resultado == int(resultado):  # Verifica si el resultado es un entero
            return int(resultado)
    except ZeroDivisionError:
        pass
    except OverflowError:
        pass
    except Exception:
        pass
    return None

# Función de fitness que mide la calidad de una solución
def fitness(expr):
    resultado = evaluar_expresion(expr)
    if resultado is not None:
        return 1 / (1 + abs(resultado))
    return 0

# Generar dígitos y operadores aleatorios
def generar_digitos_y_operadores():
    digitos = random.sample(list(map(str, range(1, 10))), 5)  # Genera 5 dígitos únicos
    operadores = ['+', '-', '*', '/']  # Los 4 operadores básicos
    return digitos, operadores

# Generar vecino cambiando un dígito o permutando operadores
def generar_vecino(digitos, operadores):
    vecino_digitos = digitos[:]
    vecino_operadores = operadores[:]
    if random.random() < 0.5:
        # Cambiar un dígito
        idx = random.randint(0, len(digitos) - 1)
        current_digits = set(digitos)
        new_digit = random.choice([d for d in map(str, range(1, 10)) if d not in current_digits])
        vecino_digitos[idx] = new_digit
    else:
        # Permutar operadores
        idx1, idx2 = random.sample(range(len(operadores)), 2)
        vecino_operadores[idx1], vecino_operadores[idx2] = vecino_operadores[idx2], vecino_operadores[idx1]
    return vecino_digitos, vecino_operadores

# Algoritmo de enfriamiento simulado
def enfriamiento_simulado():
    # Inicializar dígitos y operadores
    digitos, operadores = generar_digitos_y_operadores()
    expr = generar_expresion(digitos, operadores)
    mejor_digitos = digitos
    mejor_operadores = operadores
    mejor_expr = expr
    mejor_fitness = fitness(expr)

    # Inicializar temperatura
    temperatura = INITIAL_TEMPERATURE

    # Almacenar resultados enteros y sus expresiones
    resultados_enteros = {}

    while temperatura > FINAL_TEMPERATURE:
        for _ in range(ITERATIONS_PER_TEMP):
            # Generar vecino y evaluar
            vecino_digitos, vecino_operadores = generar_vecino(digitos, operadores)
            nueva_expr = generar_expresion(vecino_digitos, vecino_operadores)
            nueva_fitness = fitness(nueva_expr)

            # Calcular cambio en fitness
            delta_fitness = nueva_fitness - mejor_fitness

            if delta_fitness > 0 or math.exp(delta_fitness / temperatura) > random.random():
                digitos = vecino_digitos
                operadores = vecino_operadores
                expr = nueva_expr
                if nueva_fitness > mejor_fitness:
                    mejor_digitos = vecino_digitos
                    mejor_operadores = vecino_operadores
                    mejor_expr = nueva_expr
                    mejor_fitness = nueva_fitness

            # Almacenar resultado entero si existe
            resultado = evaluar_expresion(expr)
            if resultado is not None:
                resultados_enteros[resultado] = expr

        # Enfriar la temperatura
        temperatura *= ALPHA

    # Evaluar los resultados finales
    resultado = evaluar_expresion(mejor_expr)
    if resultado is not None:
        resultados_enteros[resultado] = mejor_expr

    # Determinar el valor mínimo y máximo y sus expresiones
    if resultados_enteros:
        valor_minimo = min(resultados_enteros.keys())
        valor_maximo = max(resultados_enteros.keys())
        expresion_minima = resultados_enteros[valor_minimo]
        expresion_maxima = resultados_enteros[valor_maximo]
    else:
        valor_minimo = None
        valor_maximo = None
        expresion_minima = None
        expresion_maxima = None

    print("Resultado:", resultado)
    print("Valor mínimo:", valor_minimo, "con expresión:", expresion_minima)
    print("Valor máximo:", valor_maximo, "con expresión:", expresion_maxima)
    return valor_minimo, valor_maximo, expresion_minima, expresion_maxima

# Ejecutar el algoritmo de enfriamiento simulado
enfriamiento_simulado()


Mejor expresión: 3-5*2+7/1
Resultado: 0
Valor mínimo: -69 con expresión: 1-9*8+4/2
Valor máximo: 77 con expresión: 9*8-2+7/1


(-69, 77, '1-9*8+4/2', '9*8-2+7/1')

# (*)Calcula la complejidad del algoritmo

Por la misma naturaleza del algoritmo creado vamos a utilizar la temperatura inicial, la temperatura final y la tasa de enfriamiento para calcular la complejidad del algoritmo de enfriamiento simulado.

### Número de iteraciones (N)
El cálculo del número de iteraciones del algoritmo esta directamente relacionado con cómo la temperatura se reduce desde el valor inicial hasta el valor final usando la tasa de enfriamiento. Usaremos la siguiente formula:

**Número de Iteraciones = [(Ln(T_final) - Ln(T_initial)]/Ln (α)**

Donde <br>
Temperatura Inicial (T_initial): Define el punto de partida del algoritmo. <br>
Temperatura Final (T_final): Define cuándo se detiene el algoritmo. <br>
Tasa de Enfriamiento (α): Determina la rapidez con la que la temperatura disminuye. <br>
Resolviendo: <br>
Número de Iteraciones=(0−9.21034)/−0.00501=1838.24 aproximando a **1838**
​
### Número de evaluaciones por iteración (M)
**ITERATIONS_PER_TEMP** se define en el algoritmo, en este caso es 100

******
<br>
La complejidad del algoritmo de enfriamiento simulado en términos generales es
O(N×M), donde

**N** es el número de iteraciones determinado por la tasa de enfriamiento y los límites de temperatura, y

**M** es el número de evaluaciones por iteración. En el caso específico con los parámetros dados, se espera que la complejidad sea de alrededor de 80,000 evaluaciones.

**Complejidad total = 1838 x 100 = 183 800**

Según el problema (y tenga sentido), diseña un juego de datos de entrada aleatorios

Respuesta

Aplica el algoritmo al juego de datos generado

Respuesta

# Enumera las referencias que has utilizado(si ha sido necesario) para llevar a cabo el trabajo

Para encontrar información acerca del algoritmo de enfriamiento simulado he tenido que consultar las referencias bibliográficas:

* "Material docente del curso Algoritmos de optimización" por Raúl Reyero Díez

* "Optimization by Simulated Annealing" por S. Kirkpatrick, C.D. Gelatt, y M.P. Vecchi
https://www2.stat.duke.edu/~scs/Courses/Stat376/Papers/TemperAnneal/KirkpatrickAnnealScience1983.pdf

* "Handbook of Metaheuristics" por Michel Gendreau · Jean-Yves Potvin Editors
http://old.math.nsc.ru/LBRT/k5/OR-MMF/2019_Book_HandbookOfMetaheuristics.pdf


Describe brevemente las lineas de como crees que es posible avanzar en el estudio del problema. Ten en cuenta incluso posibles variaciones del problema y/o variaciones al alza del tamaño

Respuesta